frame/staking/src/weights.rs

```diff
@@ -380,43 +407,60 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	// Storage: Staking Validators (r:501 w:0)
 	fn get_npos_targets(v: u32, ) -> Weight {
 		(0 as Weight)
-			// Standard Error: 33_000
-			.saturating_add((11_702_000 as Weight).saturating_mul(v as Weight))
+			// Standard Error: 35_000
+			.saturating_add((11_967_000 as Weight).saturating_mul(v as Weight))
 			.saturating_add(T::DbWeight::get().reads(1 as Weight))
 			.saturating_add(T::DbWeight::get().reads((1 as Weight).saturating_mul(v as Weight)))
 	}
+	// Storage: Staking MinNominatorBond (r:0 w:1)
 	// Storage: Staking MaxNominatorsCount (r:0 w:1)
 	// Storage: Staking ChillThreshold (r:0 w:1)
 	// Storage: Staking MaxValidatorsCount (r:0 w:1)
 	// Storage: Staking MinValidatorBond (r:0 w:1)
-	// Storage: Staking MinNominatorBond (r:0 w:1)
 	fn set_staking_limits() -> Weight {
-		(6_650_000 as Weight)
+		(7_080_000 as Weight)
 			.saturating_add(T::DbWeight::get().writes(5 as Weight))
 	}
-	// Storage: Staking CounterForValidators (r:1 w:1)
-	// Storage: Staking Validators (r:1 w:1)
 	// Storage: Staking MaxValidatorsCount (r:1 w:0)
+	// Storage: Staking VoterBags (r:1 w:1)
+	// Storage: Staking CounterForVoters (r:1 w:1)
+	// Storage: Staking VoterBagFor (r:1 w:1)
+	// Storage: Staking Validators (r:1 w:1)
+	// Storage: Staking CounterForValidators (r:1 w:1)
 	// Storage: Staking Ledger (r:1 w:0)
-	// Storage: Staking Nominators (r:1 w:0)
 	// Storage: Staking ChillThreshold (r:1 w:0)
+	// Storage: Staking VoterNodes (r:1 w:1)
+	// Storage: Staking Nominators (r:1 w:0)
 	// Storage: Staking MinValidatorBond (r:1 w:0)
 	fn chill_other() -> Weight {
-		(92_104_000 as Weight)
+		(93_713_000 as Weight)
 			.saturating_add(T::DbWeight::get().reads(11 as Weight))
 			.saturating_add(T::DbWeight::get().writes(6 as Weight))
 	}
+	// Storage: Staking Ledger (r:1 w:0)
+	// Storage: Staking Bonded (r:1 w:0)
+	// Storage: Staking VoterBagFor (r:1 w:1)
+	// Storage: Staking VoterBags (r:2 w:2)
```

Comment:
why are we touching two voter bags here? is it because the worse case is to assume that the person being rebagged is the head? in that case it makes sense.

Reply:
I think its because we always do `bag.put()` on the destination bag as well as the source bag. Maybe `remove_node` could return bool indicating if the head or tail of the bag was updated. If it is updated then we can do `bag.put`, otherwise we can skip the write. Regardless though, we need the 2 reads (unless we modify the nodes to know if they are a head or a tail)

```
			// clear the old bag head/tail pointers as necessary
			if let Some(mut bag) = Bag::<T>::get(node.bag_upper) {
				bag.remove_node(&node);
				bag.put();
			} else {
				debug_assert!(false, "every node must have an extant bag associated with it");
				crate::log!(
					error,
					"Node for staker {:?} did not have a bag; VoterBags is in an inconsistent state",
					node.voter.id,
				);
			}

			// put the voter into the appropriate new bag
			let new_idx = notional_bag_for::<T>(weight_of(&node.voter.id));
			node.bag_upper = new_idx;
			let mut bag = Bag::<T>::get_or_make(node.bag_upper);
			bag.insert_node(node);
			bag.put();

Reply:
indeed, we can totally prevent read/write to bags. If you rebag from non-head-tail to non-head-tail, which is arguably the majority of the cases, you do not need to touch any `Bag`. 

Not important to get done now or here, but would be great if you put a TODO in the code for it.

Reply:
> If you rebag from non-head-tail to non-head-tail

I think the insert still needs a read of the bag unless we know ahead of the time the two nodes we are inserting in between.

Also, in the current state we always insert a new node as the tail